live-build/config/hooks/configuration/download-packages.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash -eu
+
+set -o pipefail
+
+function download_packages() {
+	echo "Downloading: $*"
+	for _ in {1..5}; do
+		apt-get download "$@" && return
+		echo "Download failed, retrying."
+		sleep 5
+	done
+	echo "Failed to download packages after 5 attempts. Aborting." >&2
+	exit 1
+}
+
+mkdir -p /packages
+cd /packages
+
+apt-get update
+
+dpkg-query -Wf '${Package}=${Version}\n' >/tmp/packages
+
+PACKAGES=()
+while read -r package; do
+	PACKAGES+=("$package")
+	if [[ "${#PACKAGES[@]}" -ge 20 ]]; then
+		download_packages "${PACKAGES[@]}"
+		PACKAGES=()
+	fi
+done </tmp/packages
+
+[[ "${#PACKAGES[@]}" -ne 0 ]] && download_packages "${PACKAGES[@]}"
```

Comment:
could this be replaced with something like this:
```
export -f download_packages
dpkg-query -Wf '${Package}=${Version}\n' | xargs -n 20 bash -c 'download_packages "$@"' _
```
?

Reply:
That's much more concise. It seems like it doesn't abort if `download_packages` fails though, not sure why.

Reply:
What do you mean? Here's what I see in my testing...

- All succeed, it returns success:
```
± function dostuff() { exit 0; }
± echo 1 2 3 4 | xargs -n2 bash -c 'dostuff "$@"' _
± echo $?
0
```

- All fail, it returns failure:
```
± function dostuff() { exit 1; }
± echo 1 2 3 4 | xargs -n2 bash -c 'dostuff "$@"' _
± echo $?
123
```

- one fails, it returns failure:
```
± function dostuff() { local R=$((RANDOM % 2)); echo $R; exit $R; }
± echo 1 2 3 4 | xargs -n1 bash -c 'dostuff "$@"' _
0
1
0
0
± echo $?
123
```

Reply:
Yeah, in my quick testing I was running something very similar to your 3rd case, where we can see it doesn't stop on failure. It does return an error code at the end though, so it should be fine. I've updated the review.